File: TUnit.TestProject/AfterTestAttributeTests.cs
using TUnit.Core.Interfaces;

namespace TUnit.TestProject;

public class AfterTestAttributeTests
{
    private static readonly string Filename = $"{Guid.NewGuid():N}-AfterTestAttributeTests.txt";

    [Test]
    [WriteFileAfterTest]
    public async Task Test()
    {
        await Assert.That(File.Exists(Filename)).IsFalse();
    }

    public class WriteFileAfterTestAttribute : Attribute, ITestEndEventReceiver
    {
        public async ValueTask OnTestEnd(TestContext testContext)
        {
            Console.WriteLine(@"Writing file inside WriteFileAfterTestAttribute!");

            await FilePolyfill.WriteAllTextAsync(Filename, "Foo!");
        }

        public int Order => 0;
    }
}

File: TUnit.TestProject.FSharp/AfterTestAttributeTests.fs
namespace TUnit.TestProject.FSharp

open System
open System.IO
open System.Threading.Tasks
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Core.Interfaces
open TUnit.Core
open TUnit.Assertions.FSharp.Operations


type WriteFileAfterTestAttribute() =
        inherit System.Attribute()
        interface ITestEndEventReceiver with
            member _.OnTestEnd(testContext: AfterTestContext) =
                task {
                    Console.WriteLine("Writing file inside WriteFileAfterTestAttribute!")
                    do! FilePolyfill.WriteAllTextAsync(filename, "Foo!")
                } :> Task
            member _.Order = 0

type AfterTestAttributeTests() =
    static let filename = sprintf "%s-AfterTestAttributeTests.txt" (Guid.NewGuid().ToString("N"))

    [<Test>]
    [<WriteFileAfterTest>]
    member _.Test() =
        async{
             do! check(Assert.That<bool>(File.Exists(filename)).IsFalse())
        }
